web-console/src/lib/services/pipelineManagerQuery.ts

```diff
@@ -63,62 +66,117 @@ const consolidatePipelineStatus = (newPipeline: Pipeline, oldPipeline: Pipeline
+  pipelines: () =>
+    PipelinesService.listPipelines().then(ps =>
+      ps.map(p => ({
+        ...p,
+        state: { ...p.state, current_status: toClientPipelineStatus(p.state.current_status) }
+      }))
     ),
-    pipelineConfig: PipelinesService.getPipelineConfig,
-    pipelineStats: PipelinesService.pipelineStats,
-    pipelineLastRevision: PipelinesService.pipelineDeployed,
-    pipelineValidate: PipelinesService.pipelineValidate,
-    connectors: () => ConnectorsService.listConnectors(),
-    connectorStatus: ConnectorsService.getConnector,
-    getAuthConfig: AuthenticationService.getAuthenticationConfig,
-    listApiKeys: ApiKeysService.listApiKeys,
-    getApiKey: ApiKeysService.getApiKey,
-    listServices: ServicesService.listServices,
-    getService: ServicesService.getService,
-    newService: ServicesService.newService
-  })
-)
+  pipelineStatus: compose(PipelinesService.getPipeline, p =>
+    p.then(p => ({ ...p, state: { ...p.state, current_status: toClientPipelineStatus(p.state.current_status) } }))
+  ),
+  pipelineConfig: PipelinesService.getPipelineConfig,
+  pipelineStats: PipelinesService.pipelineStats,
+  pipelineLastRevision: PipelinesService.pipelineDeployed,
+  pipelineValidate: PipelinesService.pipelineValidate,
+  connectors: () => ConnectorsService.listConnectors(),
+  connectorStatus: ConnectorsService.getConnector,
+  listApiKeys: ApiKeysService.listApiKeys,
+  getApiKey: ApiKeysService.getApiKey,
+  listServices: ServicesService.listServices,
+  getService: ServicesService.getService,
+  newService: ServicesService.newService
+}
+
+export const makePipelineManagerQuery = ({
+  onError = defaultPipelineManagerQueryOptions.onError
+}: PipelineManagerQueryOptions) =>
+  (({ pipelines, pipelineStatus, ...queries }) => ({
+    ...queries,
+    pipelines: () => ({
+      ...pipelines(),
+      // Avoid displaying PROVISIONING status when local status is more detailed
+      structuralSharing<T>(oldData: T | undefined, newData: T) {
+        invariant(((data: any): data is Pipeline[] | undefined => true)(oldData))
+        invariant(((data: any): data is Pipeline[] => true)(newData))
+        return leftJoin(
+          newData,
+          oldData ?? [],
+          p => p.descriptor.pipeline_id,
+          p => p.descriptor.pipeline_id,
+          consolidatePipelineStatus
+        ) as T
+      }
+    }),
+    pipelineStatus: (pipelineName: string) => ({
+      ...pipelineStatus(pipelineName),
+      structuralSharing<T>(oldData: T | undefined, newData: T) {
+        invariant(((data: any): data is Pipeline | undefined => true)(oldData))
+        invariant(((data: any): data is Pipeline => true)(newData))
+        return consolidatePipelineStatus(newData, oldData) as T
+      }
+    })
+  }))(
+    mkQuery(PipelineManagerApi, {
+      wrapQueryFn: f =>
+        ((...args: Arguments<typeof f>) => {
+          // A single retry is performed
+          // Refactor to support arbitrary number of retries
```

Comment:
is this a TODO?
It should be marked TODO.

Reply:
It is not a TODO yet before there was not and might not be a requirement to update this code; can mark as a TODO if you think it's better